File: src/structural/proxy/typescript/classes/UserProxy.ts
import { AdminUser } from './AdminUser';
import { UserAddressProtocol } from './interfaces/UserAddressProtocol';
import { UserProtocol } from './interfaces/UserProtocol';

export class UserProxy extends UserProtocol {
  private realUser: AdminUser | null = null;
  private realUserAddress: UserAddressProtocol[] | null = null;

  private createUser(): AdminUser {
    return new AdminUser(this.firstName, this.userName);
  }

  async getAddresses(): Promise<UserAddressProtocol[]> {
    if (!this.realUser) this.realUser = this.createUser();
    if (!this.realUserAddress)
      this.realUserAddress = await this.realUser.getAddresses();

    return this.realUserAddress;
  }
}

File: src/structural/proxy/python/classes/user_proxy.py
from typing import List

from classes.admin_user import AdminUser
from classes.interfaces.user_address_protocol import UserAddressProtocol
from classes.interfaces.user_protocol import UserProtocol


class UserProxy(UserProtocol):
    __real_user: AdminUser | None = None
    __real_user_address: List[UserAddressProtocol] | None = None

    def create_user(self) -> AdminUser:
        return AdminUser(self.first_name, self.user_name)

    async def get_addresses(self) -> List[UserAddressProtocol]:
        if self.__real_user is None:
            self.__real_user = self.create_user()

        if self.__real_user_address is None:
            self.__real_user_address = await self.__real_user.get_addresses()

        return self.__real_user_address
